site/content/languages/zig.md

```diff
@@ -0,0 +1,213 @@
+calls a free function, with the same `ptr`, afterwards and unconditionally.
+
+Note: wazero includes an [example project][9] that shows this.
+
+The [zig example][9] does a few things of interest:
+* Uses `@ptrToInt` to change a Zig pointer to a numeric type
+* Uses `[*]u8` as an argument to take a pointer and slices it to build back a
+string
+* It also shows how to import a host function using the `extern` directive
+
+To allow the host to allocate memory, you need to define your own `malloc` and
+`free` functions:
+```webassembly
+(func (export "malloc") (param $size i32) (result (;$ptr;) i32))
+(func (export "free") (param $ptr i32) (param $size i32))
+```
+
+Because Zig easily allows end-users to [plug their own allocators][12], it relatively easy to 
+export custom `malloc`/`free` pairs to the host.
+
+For instance, the following code exports `malloc`, `free` from Zig's `page_allocator`:
+
+```zig
+const allocator = std.heap.page_allocator;
+
+pub export fn malloc(length: usize) ?[*]u8 {
+    const buff = allocator.alloc(u8, length) catch return null;
+    return buff.ptr;
+}
+
+pub export fn free(buf: [*]u8, length: usize) void {
+    allocator.free(buf[0..length]);
+}
+```
+
+## System Calls
+
+Please read our overview of WebAssembly and
+[System Calls]({{< ref "_index.md#system-calls" >}}). In short, WebAssembly is
+a stack-based virtual machine specification, so operates at a lower level than
+an operating system.
+
+For functionality the operating system would otherwise provide, you must use
+the `wasm32-wasi` target. This imports host functions in
+[WASI]({{< ref "/specs#wasi" >}}).
+
+For example, `zig build-exe -target wasm32-wasi hello.zig` compiles the
+below `main` function into a WASI function exported as `_start`.
+
+```zig
+const std = @import("std");
+const io = std.io;
+const os = std.os;
+const allocator = std.heap.page_allocator;
+const stdout = std.io.getStdOut();
+```
+
+Note: wazero includes an [example WASI project][10] including [source code][11]
+that implements `cat` without any WebAssembly-specific code.
+
+## Concurrency
+
+Please read our overview of WebAssembly and
+[concurrency]({{< ref "_index.md#concurrency" >}}). In short, the current
+WebAssembly specification does not support parallel processing.
+
+## Optimizations
+
+Below are some commonly used configurations that allow optimizing for size or
+performance vs defaults. Note that sometimes one sacrifices the other.
+
+### Binary size
+
+Those with `%.wasm` binary size constraints can change their source, 
+e.g. picking a [different allocator][9b] or set `zig` flags to reduce it.
+
+Source changes:
+* [WebAssembly-only fast allocator][12] is a specialized allocator that 
+trades generality for simplicity, making however specific assumptions 
+on the runtime environment (e.g. single-threaded access to global variables, 
+it direcly calls Wasm-specific intrinsics and it never releases memory)
```

Comment:
I see this is still on `main`, so I am assuming this has not been folded back into the GPA yet (although the source says "//! This is intended to be merged into GeneralPurposeAllocator at some point.")

Reply:
If this isn't released or widely used, yet, we maybe don't need to add it. Are we using it in examples?

This is a first-time doc so we probably should hint to well used/tested things.

Reply:
no we are not using it in examples and -- I have just checked -- it has not been released yet, so let's skip it for now